### DISCUSSION TITLE
Novice to VS/c++: help getting project to compile

### DISCUSSION THREAD

**Discussion by RablinQ** [jovibor/HexCtrl#94] [Q&A] [answered]
I'm trying hard to get HexCtrl into my own project and getting it compiled.
The list of errors gets smaller but in the end, whatever I try to do, the project produces errors which do not give hints how to solve it.

One of the reasons I have issues is that there are tons of settings on project level, linker settings, compiler settings, path settings etc which makes this a hell to get it working.

I hope to get a step-by-step explanation for adding HexCtrl into an existing project. Maybe other novice users can be helped by this.

So first I tried to use the sample project given in HexCtrl to be sure that one builds and works:

HexCtrl\VS Projects\HexCtrl.sln

Using VS2020 (having latest updates) this works out-of-the-box.
I'm interested in the project **Sample Dialog** as this is what I need: my own project (a dll) should be able to HexEdit in a dialog where I'm not using the HexCtrl dll as reference but include HexCtrl source into my project (so no external references needed).

My own project works in VS2020, ISO C++20 Standard (/std:c++20) and has Use MFC in a Static Library.

Now reading the docs of HexCtrl:
Integrate Sources

> To build HexCtrl from the sources:
> 
>     Add all files from the HexCtrl folder into your project
>     (can skip adding rapidjson-amalgam.h)

This is already too vague for me. What I did is copy the folder to my project folder and drag-drop the folder ontpo my project in VS2020. All files are added.

I first need my project buildable only having done this. So not using any features of HexCtrl, just place the folder into my project and get this buildable.

When building I get these messages:

```
Rebuild started at 13:51...
1>------ Rebuild All started: Project: MyDll, Configuration: Release Win32 ------
1>ListEx.ixx
1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winnt.h(1005,1): message : sorry: 'inline' function definition for 'ULONGLONG Int64ShllMod32(ULONGLONG,DWORD)' could not be written to module due to unrecognized expression or statement at 'C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winnt.h(1010,5)'.
1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winnt.h(1005,1): message : If possible please provide a repro here: https://developercommunity.visualstudio.com
1>(compiling source file 'HexCtrl/dep/ListEx/ListEx.ixx')
1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winnt.h(1021,1): message : sorry: 'inline' function definition for 'LONGLONG Int64ShraMod32(LONGLONG,DWORD)' could not be written to module due to unrecognized expression or statement at 'C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winnt.h(1026,5)'.
1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winnt.h(1021,1): message : If possible please provide a repro here: https://developercommunity.visualstudio.com
```

Which is abracadabra to me and no info I can find on this. Is this a blocking error?

After that multiple errors like:

```
Cannot open include file: 'stdafx.h': No such file or directory
Cannot open include file: 'stdafx.h': No such file or directory
Cannot open include file: 'stdafx.h': No such file or directory
Cannot open include file: 'stdafx.h': No such file or directory
Cannot open include file: 'stdafx.h': No such file or directory
Cannot open include file: 'stdafx.h': No such file or directory
Cannot open include file: 'stdafx.h': No such file or directory
Cannot open include file: 'stdafx.h': No such file or directory
```

I do not have this file indeed and HexCtrl folder does not contain it.

So I used the same file which is present in the **Sample Dialog** project.
Copied stdafx.h and targetver.h from there into my project and added it.
I had to change multiple HexCtrl cpp files so each file can find it.

Now the list of errors has reduced to:

```
_DllMain@12 already defined in MyDll.obj
one or more multiply defined symbols found
```

Searching for DllMain I only get the one method defined in my own project.
So the linker has issues?

What can I check so that this works?

**Comment by jovibor** [comment, marked as answer]
Hi,

Indeed HexCtrl is a bit complex project.
```cpp
message : sorry: 'inline' function definition for 'ULONGLONG Int64ShllMod32(ULONGLONG,DWORD)' could not be written to module due to unrecognized expression or statement at 'C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winnt.h(1010,5)'.
```
You can ignore these messages all together. It's just the compiler can't handle old stuff with new modules` model, it'll be fixed (I hope) in future VS versions.

`Cannot open include file: 'stdafx.h': No such file or directory`
This one is pretty standard stuff. HexCtrl uses `stdafx.h` for precompiled headers (until it completely moves to modules).
You can check it in your project settings:
![image](https://github.com/user-attachments/assets/2fa331f5-8f3f-4471-bd88-1d0a22142ede)
The name of Precompile Header File must coincide within your project and what HexCtrl uses (stdafx.h). If they differs you can change one or another, throughout your project or within the HexCtrl.

Also, precompiled headers do not work for modules. So you must disable this feature for all `*.ixx` files.
Right click on `*.ixx` file in the Solution Explorer and make sure that Precompiled header option is set to **Not Using Precompiled Headers**:
![image](https://github.com/user-attachments/assets/41946a53-8841-4420-9fd1-db7ac0f7095c)

> So I used the same file which is present in the Sample Dialog project.
> Copied stdafx.h and targetver.h from there into my project and added it.

Don't do that, these files are only for Sample Project, your project has its own stuff.
All you need is exactly in the `HexCtrl/HexCtrl` subfolder as stated in the docs.

If you have some questions left feel free to ask.

**Comment by RablinQ** [comment]
Thanks for your help. I'm happy !

I got my project buildable now.
For others having the same issue I explain what I  encountered and how to solve it.

I have stdafx.h placed in the root of my project.
HexCtrl cpp files are spread over subfolders within folder HexCtrl placed in the root.

For those cpp files to find stdafx.h I needed to include the file like this:

#include "../../stdafx.h"

The cpp file can find it, no errors at design time.
But at compile time I get this error:

`unexpected end of file while looking for precompiled header. Did you forget to add '#include "stdafx.h"' to your source?`

for each cpp file which has relative include path to that file.

Seems that the include needs to be consistent for each cpp file to get rid of this error:

Add the root folder containing stdafx.h to your project’s include directories in the Visual Studio project settings. This ensures that the relative path is no longer necessary, and you can use:

`#include "stdafx.h"`

Add this to the project settings for Additional Include Directories:

`$(ProjectDir)`

Then all cpp files can access stdafx.h without using relative paths.


Another issue I encountered. My project was initially a non MFC project.
Now it is using MFC, my custom DllMain conflicts with the one provided with MFC.
This results in error:

`_DllMain@12 already defined error`

I solved this by creating a (static) class which has a constructor/destructor which executes the same code as I have in my DllMain method.
(actions to init and cleanup). I could then remove my DllMain method.

Now the project builds and works correctly in my application.

Now the next step is to use HexCtrl to add new feature to my application.

Thanks again for helping !!!
